fem/fespace.hpp

```diff
@@ -526,10 +528,17 @@ class FiniteElementSpace
        is preserved. */
    void ReorderElementToDofTable();
 
+   /** @brief  Generates partial face_dof table.
+
+       The table is only defined for exterior faces that coincide with a boundary.
+       The routine uses the bdrElem_dof table and the mesh boundary information.*/
+   void GenerateFaceDofsFromBdr();
+
    void BuildDofToArrays();
 
    const Table &GetElementToDofTable() const { return *elem_dof; }
-   const Table &GetBdrElementToDofTable() const { return *bdrElem_dof; }
+   const Table *GetBdrElementToDofTable() const { return bdrElem_dof; }
+   const Table *GetFaceToDofTable() const { return face_dof; }
```

Comment:
I don't see a good reason for changing the API for `GetBdrElementToDofTable()`. I know that right now the pointer can be NULL but that is a temporary state (we can just error out in that case) -- in the future the Table can be generated on-demand. Ultimately, we will need this Table, e.g. for PA on boundaries.

Similar logic applies to the `face_dof` Table -- can be generated if requested.

Reply:
I agree that generating the tables on demand could be a good solution, and probably better than changing the API (that being said, I believe this function is extremely rarely if ever used in user code).

Erroring out is also fine, except it gives the caller no way to recover, and there isn't a way to query whether the tables are non-NULL.

Reply:
Lazy generation for the non-NURBS case is trivial in both cases (very similar to `FiniteElementSpace::BuildElementToDofTable`). I can add it, if you guys are fine with that.

Reply:
I am fine with that.